### PR TITLE
[DRAFT] Support sample weights for retrieval models trained with contrastive task

### LINKED ISSUE #789: [FEA] Easy functionality for using class_weights and sample_weights for Retrieval models
# 🚀 Feature request

- Since Inbatch sampler converts the retrieval task to multi-class classification,  we need an easy functionality to be able to use `class_weight` and `sample_weight` args in the model.fit() for retrieval models, without users to deal with customization.
- add unit test for using `sample_weight` and `class_weight` args for retrieval tasks in [here](https://github.com/NVIDIA-Merlin/models/blob/main/tests/unit/tf/models/test_retrieval.py)
- we need examples or code snippets in the docs to showcase the functionality

This request comes from a customer.

### PR DESCRIPTION
**Pull request by sararb**
try to fix #789

This draft PR shows an example of setting sample weights for positive and sampled negative candidates.

### CODE REVIEW

File: merlin/models/tf/transforms/bias.py [hunk 1]
```diff
@@ -300,3 +300,42 @@ def from_config(cls, config):
         config["item_freq_probs"] = tf.convert_to_tensor(config["item_freq_probs"])
 
         return cls(**config)
+
+
+class RetrievalSampleWeight(Block):
+    def __init__(self, pos_class_weight, neg_class_weight=1, **kwargs):
+        """Generate the sample weights for positive and sampled
+        negative candiddates
```
Comment (gabrielspmoreira) [inline merlin/models/tf/transforms/bias.py:308]:
```suggestion
        negative candidates
```

File: merlin/models/tf/transforms/bias.py [hunk 2]
```diff
@@ -300,3 +300,42 @@ def from_config(cls, config):
...
+        super().__init__(**kwargs)
+
+    def call(
+        self,
+        outputs: Prediction,
+        training=False,
+        testing=False,
+        **kwargs,
+    ) -> Prediction:
+        """Update the predictins returned by the
```
Comment (gabrielspmoreira) [inline merlin/models/tf/transforms/bias.py:328]:
```suggestion
        """Update the predictions returned by the
```

File: merlin/models/tf/transforms/bias.py [hunk 3]
```diff
@@ -300,3 +300,42 @@ def from_config(cls, config):
...
+        for negative and positive candidates.
+        """
+        if training:
+            predictions = outputs.predictions
+            bs, num_samples = tf.shape(predictions)
+            # generate a 2-d matrix of sample weights
+            neg_samples = tf.ones((bs, num_samples - 1)) * self.neg_class_weight
+            pos_samples = tf.ones((bs, 1)) * self.pos_class_weight
+            samples_weights = tf.concat([pos_samples, neg_samples], axis=1)
+            samples_weights = tf.expand_dims(samples_weights, -1)
```
Comment (gabrielspmoreira) [inline merlin/models/tf/transforms/bias.py:339]: Does it become a 3D matrix (batch size, 1 + #negatives, 1) after the `expand_dims`? Does it work with categorical_cross_entropy?

File: tests/unit/tf/models/test_retrieval.py [hunk 4]
```diff
@@ -515,6 +515,36 @@ def test_two_tower_model_with_custom_options(
     )
 
 
+def test_two_tower_v2_with_sample_weight(ecommerce_data: Dataset):
```
Comment (gabrielspmoreira) [inline tests/unit/tf/models/test_retrieval.py:518]: I think it would be good using the binary cross entropy loss in the test, so that it serve as an example

**Review by gabrielspmoreira — COMMENTED** [review]
As we discussed, I think it should be good to include this under a new `experimental` package

**Comment by sararb** [comment]
closing this PR as a new one is opened [here](https://github.com/NVIDIA-Merlin/models/pull/946/files) and includes an updated version of the SampleWeight block
